e2e/internal/e2e/profile/profile.go

```diff
@@ -0,0 +1,230 @@
+
+import (
+	"strings"
+	"testing"
+
+	"github.com/sylabs/singularity/e2e/internal/e2e"
+	"github.com/sylabs/singularity/internal/pkg/fakeroot"
+	"github.com/sylabs/singularity/internal/pkg/test/tool/require"
+	"github.com/sylabs/singularity/internal/pkg/util/user"
+)
+
+// Profile represents various properties required to run an E2E test
+// under a particular user profile.
+type Profile interface {
+	// Privileged returns whether the test should be executed with
+	// elevated privileges or not.
+	Privileged() bool
+
+	// Requirements calls the different require.* functions
+	// necessary for running an E2E test under this profile.
+	Requirements(t *testing.T)
+
+	// Args returns the additional arguments, if any, to be passed
+	// to the singularity command specified by cmd in order to run a
+	// test under this profile.
+	Args(cmd []string) []string
+
+	// User returns the user to be used during the execution of the
+	// singularity command.
+	User(t *testing.T) *user.User
+
+	// In returns true if the specified list of profiles contains
+	// this profile.
+	In(profiles ...Profile) bool
+
+	// String provides a string representation of this profile.
+	String() string
+}
+
+// baseProfile almost implements the Profile interface (but it doesn't,
+// it's missing the In and String methods). This exists in order to
+// reduce code duplication.
+type baseProfile struct{}
+
+func (baseProfile) Privileged() bool {
+	return false
+}
+
+func (baseProfile) Requirements(t *testing.T) {
+}
+
+func (baseProfile) Args([]string) []string {
+	return nil
+}
+
+func (baseProfile) User(t *testing.T) *user.User {
+	return getOrigUser(t)
+}
+
+// User implements the Profile interface describing a non-privileged
+// user.
+type User struct {
+	baseProfile
+}
+
+func (p User) In(profiles ...Profile) bool {
+	return in(p, profiles)
+}
+
+func (User) String() string {
+	return "User"
+}
+
+// Root implements the Profile interface describing a privileged user.
+type Root struct {
+	baseProfile
+}
+
+func (Root) Privileged() bool {
+	return true
+}
+
+func (Root) User(t *testing.T) *user.User {
+	u, err := user.GetPwUID(0)
+	if err != nil {
+		t.Fatalf("failed to retrieve root user information: %s", err)
+	}
+	return u
+}
+
+func (p Root) In(profiles ...Profile) bool {
+	return in(p, profiles)
+}
+
+func (Root) String() string {
+	return "Root"
+}
+
+// Fakeroot implements the Profile interface describing a "fake" root
+// user and makes use of singularity --fakeroot flag where appropriate.
+type Fakeroot struct {
+	baseProfile
+}
+
+func (Fakeroot) Requirements(t *testing.T) {
+	require.UserNamespace(t)
+
+	// check that current user has valid mappings in /etc/subuid
+	if _, err := fakeroot.GetIDRange(fakeroot.SubUIDFile, uint32(e2e.OrigUID())); err != nil {
+		t.Fatalf("fakeroot configuration error: %s", err)
+	}
+
+	// check that current user has valid mappings in /etc/subgid
+	if _, err := fakeroot.GetIDRange(fakeroot.SubGIDFile, uint32(e2e.OrigUID())); err != nil {
```

Comment:
Shouldn't you use e2e.OrigGID() here?

Reply:
I had the exact same question! I'm glad I'm not alone :-)

The file that this is reading, the one specified by the first argument, is indexed by user name, not group name. I remember my original feedback when this function was added was that it should be two separate functions, not one function doing two different things, but some implementation details make it so that having a single function is more convenient.

I guess I can add a comment clarifying. Thanks.

Reply:
I see now what you mean after your description. Thanks.